### DISCUSSION TITLE
Refactor code to have a gamestate struct

### DISCUSSION THREAD

**Discussion by ValorZard** [ikemen-engine/Ikemen-GO#872] [Feature Request] [answered]
This is important if rollback support is going to be added, but it also allows for better training mode options as well.
There is already a proof of concept over here: https://streamable.com/15idpp with the codebase over here: https://github.com/BwdYeti/Ikemen-GO

However, here's the current status of that: 

> progress so far, setting hotkeys for saving and loading the game state, and playing around with them during an AI match
> moved a bunch of gameplay relevant data from System to a new GameState struct, which is what actually gets copied; still a bunch that would be left to do though
> also rng still isn't consistent for some reason; if it were, the AI would do the same thing every time it loads when the rng seed is loaded
Random() must be getting called from somewhere outside of gameplay stuff that I need to figure out
re: rng consistency, I didn't research it but my best guess is somewhere else in the code or the lua or something the Random function is getting called, from outside of the main gameplay loop, which is polluting the seed
if not that, it's possible the execution of parts of the code/character scripts isn't always in the same order???
though that's less likely now that I think about it, because it would suggest that when you reload the save state you'd get a few different variations you'd cycle between, instead of infinite minor permutations like what seems to be happening
https://github.com/BwdYeti/Ikemen-GO/tree/savestate
main branch is edits to separate out the state and make saving easier, savestate branch actually has the savestate functionality
there's an update to the script files too that you'll have to copy to the game directory so you can actually do the thing, F7 save F8 load
will have to look into embedding interfaces I think, because it'll be the easiest way to hash the game state without it turning into a nightmare

It would be cool if we could get more eyes on this so it can be finished up.

**Comment by K4thos** [comment]
(speaking only for myself) If @BwdYeti makes pull request tested against main branch (gacel's repo) than I'm personally ok with it but not before upcoming 0.98 release due to risk of regressions.

**Comment by ValorZard** [comment]
Quick note about this, I seem to remember @BwdYeti saying that the savestate branch is the one with saving and loading states actually working

**Comment by K4thos** [comment, marked as answer]
this is already part of nightly
